Ocpp: StatusNotificationRequest.Timestamp is optional

```
[main  ] INFO 2022/07/17 13:54:38 evcc 0.96
[main  ] INFO 2022/07/17 13:54:38 using config file c:\evcc\evcc.yaml
[main  ] INFO 2022/07/17 13:54:38 listening at :7070
[ocpp-cp] TRACE 2022/07/17 13:54:47 *core.BootNotificationRequest: &{ChargeBoxSerialNumber:ABBTerraAC ChargePointModel:CDT_TACW22::NET_WIFI ChargePointSerialNumber: ChargePointVendor:Chargedot FirmwareVersion:TAC3Z9119006710273::V1.5.2 Iccid: Imsi: MeterSerialNumber: MeterType:V1}
[ocpp-cp] TRACE 2022/07/17 13:54:48 *core.StatusNotificationRequest: &{ConnectorId:0 ErrorCode:NoError Info:null Status:Available Timestamp:<nil> VendorId: VendorErrorCode:0x00}
[ocpp  ] TRACE 2022/07/17 13:54:48 enqueued request 882357421 - GetConfiguration for client ABBTerraAC
[ocpp-cp] TRACE 2022/07/17 13:54:49 *core.StatusNotificationRequest: &{ConnectorId:1 ErrorCode:NoError Info:null Status:Available Timestamp:<nil> VendorId: VendorErrorCode:0x00}
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x1b4c7e0]

goroutine 111 [running]:
github.com/evcc-io/evcc/charger/ocpp.(*CP).StatusNotification(0xc00010edc0, 0xc0005a8cc0)
        /home/runner/work/evcc/evcc/charger/ocpp/cp_core.go:55 +0x160
github.com/evcc-io/evcc/charger/ocpp.(*CS).OnStatusNotification(0x0?, {0xc000a2a815?, 0xc00088db80?}, 0x0?)
        /home/runner/work/evcc/evcc/charger/ocpp/cs_core.go:83 +0x45
github.com/lorenzodonini/ocpp-go/ocpp1%2e6.(*centralSystem).handleIncomingRequest.func1()
        /home/runner/go/pkg/mod/github.com/lorenzodonini/ocpp-go@v0.15.0/ocpp1.6/central_system.go:507 +0x6b4
created by github.com/lorenzodonini/ocpp-go/ocpp1%2e6.(*centralSystem).handleIncomingRequest
        /home/runner/go/pkg/mod/github.com/lorenzodonini/ocpp-go@v0.15.0/ocpp1.6/central_system.go:490 +0x5fe

c:\evcc>evcc charger --log trace
[main  ] INFO 2022/07/17 13:56:59 evcc 0.96
[main  ] INFO 2022/07/17 13:56:59 using config file c:\evcc\evcc.yaml
[main  ] FATAL 2022/07/17 13:59:00 cannot create charger 'my_charger': cannot create charger 'ocpp': timeout
```

_Originally posted by @kfussmann in https://github.com/evcc-io/evcc/discussions/3806#discussioncomment-3205792_

--- a/charger/ocpp/cp.go
+++ b/charger/ocpp/cp.go
@@ -60,11 +60,11 @@ type CP struct {
 	boot        *core.BootNotificationRequest
 	status      *core.StatusNotificationRequest
 
-	meterSupported      bool
-	meterUpdated        time.Time
-	measureDoneCh       chan struct{}
-	measurements        map[string]types.SampledValue
-	meterTrickerRunning bool
+	meterSupported     bool
+	measureDoneCh      chan struct{}
+	meterUpdated       time.Time
+	measurements       map[string]types.SampledValue
+	meterTickerRunning bool
 
 	supportedNumberOfConnectors int
 	smartChargingCapabilities   smartChargingProfile

--- a/charger/ocpp/cp_core.go
+++ b/charger/ocpp/cp_core.go
@@ -9,6 +9,11 @@ import (
 	"github.com/lorenzodonini/ocpp-go/ocpp1.6/types"
 )
 
+const (
+	messageExpiry     = 30 * time.Second
+	transactionExpiry = time.Hour
+)
+
 func (cp *CP) Authorize(request *core.AuthorizeRequest) (*core.AuthorizeConfirmation, error) {
 	cp.log.TRACE.Printf("%T: %+v", request, request)
 
@@ -44,10 +49,8 @@ func (cp *CP) BootNotification(request *core.BootNotificationRequest) (*core.Boo
 
 // timestampValid returns false if status timestamps are outdated
 func (cp *CP) timestampValid(t time.Time) bool {
-	const statusExpiry = 30 * time.Second
-
 	// reject if expired
-	if time.Since(t) > statusExpiry {
+	if time.Since(t) > messageExpiry {
 		return false
 	}
 
@@ -162,7 +165,7 @@ func (cp *CP) StartTransaction(request *core.StartTransactionRequest) (*core.Sta
 
 	// create new transaction
 	if request != nil {
-		if request.Timestamp.After(time.Now().Add(-1 * time.Hour)) { // only respect transactions in the last hour
+		if time.Since(request.Timestamp.Time) < transactionExpiry { // only respect transactions in the last hour
 			cp.mu.Lock()
 			cp.currentTransaction = NewTransaction(cp.currentTransaction.ID+1, request.IdTag, request.Timestamp.Time, request.MeterStart)
 
@@ -170,7 +173,7 @@ func (cp *CP) StartTransaction(request *core.StartTransactionRequest) (*core.Sta
 
 			res.TransactionId = cp.currentTransaction.ID
 
-			if request.Timestamp.After(time.Now().Add(-30*time.Second)) && cp.meterSupported && !cp.meterTickerRunning {
+			if cp.meterSupported && !cp.meterTickerRunning && time.Since(request.Timestamp.Time) < messageExpiry {
 				go func() {
 					cp.log.TRACE.Printf("starting meter value ticker")
 					cp.meterTickerRunning = true

--- a/charger/ocpp/cp_core.go
+++ b/charger/ocpp/cp_core.go
@@ -42,6 +42,11 @@ func (cp *CP) BootNotification(request *core.BootNotificationRequest) (*core.Boo
 	return res, nil
 }
 
+func (cp *CP) timestampValid(t time.Time) bool {
+	const statusExpiry = 30 * time.Second
+	return !t.Before(cp.status.Timestamp.Time) && time.Since(t) <= statusExpiry
+}
+
 func (cp *CP) StatusNotification(request *core.StatusNotificationRequest) (*core.StatusNotificationConfirmation, error) {
 	cp.log.TRACE.Printf("%T: %+v", request, request)
 
@@ -52,7 +57,7 @@ func (cp *CP) StatusNotification(request *core.StatusNotificationRequest) (*core
 		if cp.status == nil {
 			cp.status = request
 			cp.initialized.Broadcast()
-		} else if (request.Timestamp.Equal(cp.status.Timestamp.Time) || request.Timestamp.After(cp.status.Timestamp.Time)) && request.Timestamp.After(time.Now().Add(-30*time.Second)) {
+		} else if request.Timestamp == nil || cp.timestampValid(request.Timestamp.Time) {
 			cp.status = request
 		} else {
 			cp.log.TRACE.Printf("ignoring status: %s < %s", request.Timestamp.Time, cp.status.Timestamp)

--- a/charger/ocpp/cp_core.go
+++ b/charger/ocpp/cp_core.go
@@ -44,7 +44,19 @@ func (cp *CP) BootNotification(request *core.BootNotificationRequest) (*core.Boo
 
 func (cp *CP) timestampValid(t time.Time) bool {
 	const statusExpiry = 30 * time.Second
-	return !t.Before(cp.status.Timestamp.Time) && time.Since(t) <= statusExpiry
+
+	// reject if expired
+	if time.Since(t) > statusExpiry {
+		return false
+	}
+
+	// assume having a timestamp is better than not
+	if cp.status.Timestamp == nil {
+		return true
+	}
+
+	// reject older values than we already have
+	return !t.Before(cp.status.Timestamp.Time)
 }
 
 func (cp *CP) StatusNotification(request *core.StatusNotificationRequest) (*core.StatusNotificationConfirmation, error) {

--- a/charger/ocpp/cp_core.go
+++ b/charger/ocpp/cp_core.go
@@ -104,7 +104,7 @@ func (cp *CP) Heartbeat(request *core.HeartbeatRequest) (*core.HeartbeatConfirma
 		CurrentTime: types.NewDateTime(time.Now()),
 	}
 
-	if !cp.meterTrickerRunning && cp.meterSupported {
+	if !cp.meterTickerRunning && cp.meterSupported {
 		Instance().TriggerMeterValueRequest(cp)
 	}
 
@@ -170,10 +170,10 @@ func (cp *CP) StartTransaction(request *core.StartTransactionRequest) (*core.Sta
 
 			res.TransactionId = cp.currentTransaction.ID
 
-			if request.Timestamp.After(time.Now().Add(-30*time.Second)) && cp.meterSupported && !cp.meterTrickerRunning {
+			if request.Timestamp.After(time.Now().Add(-30*time.Second)) && cp.meterSupported && !cp.meterTickerRunning {
 				go func() {
 					cp.log.TRACE.Printf("starting meter value ticker")
-					cp.meterTrickerRunning = true
+					cp.meterTickerRunning = true
 					cp.measureDoneCh = make(chan struct{})
 					ticker := time.NewTicker(15 * time.Second)
 
@@ -184,7 +184,7 @@ func (cp *CP) StartTransaction(request *core.StartTransactionRequest) (*core.Sta
 							Instance().TriggerMeterValueRequest(cp)
 						case <-cp.measureDoneCh:
 							cp.log.TRACE.Printf("returning from meter value requests")
-							cp.meterTrickerRunning = false
+							cp.meterTickerRunning = false
 							return
 						}
 					}
@@ -220,7 +220,7 @@ func (cp *CP) StopTransaction(request *core.StopTransactionRequest) (*core.StopT
 	}
 
 	if cp.meterSupported {
-		if cp.meterTrickerRunning {
+		if cp.meterTickerRunning {
 			cp.measureDoneCh <- struct{}{}
 		}
 

--- a/charger/ocpp/cp_core.go
+++ b/charger/ocpp/cp_core.go
@@ -42,6 +42,7 @@ func (cp *CP) BootNotification(request *core.BootNotificationRequest) (*core.Boo
 	return res, nil
 }
 
+// timestampValid returns false if status timestamps are outdated
 func (cp *CP) timestampValid(t time.Time) bool {
 	const statusExpiry = 30 * time.Second
 

